### PR TITLE
Feature/auth middleware

### PR DESCRIPTION
**Pull request by Sandeep1908**
**Description**
This pull request enhances the authentication middleware by adding comprehensive documentation using JSDoc comments. Documentation is an essential aspect of maintaining and understanding codebases, especially for shared modules like middleware, as it facilitates easier usage and maintenance for other developers.

### CODE REVIEW

File: middlewares/auth.js [hunk 1]
```diff
@@ -1,6 +1,35 @@
...
-    
+/**
+ * Middleware for user authentication using JWT token
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ * @param {Function} next - Express next function
+ */
+
+export const auth = async (req, res, next) => {
+  // TODO: implement the auth middleware here
```
Comment (pavankpdev) [inline middlewares/auth.js]: remove the TODO comment

File: middlewares/auth.js [hunk 2]
```diff
@@ -1,6 +1,35 @@
...
+
+    const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
+
+    const user = await findUserById(decodedToken?.userId);
+
+    if (!user) {
+      throw new ApiError(401, "Invalid Access Token");
+    }
+
+    req.user = user;
```
Comment (pavankpdev) [inline middlewares/auth.js]: adding user objects directly from the DB exposes a lot of sensitive data that is not required, such as passwords and others. Let's assign the user id to `req.user`

File: middlewares/auth.js [hunk 3]
```diff
@@ -1,6 +1,35 @@
...
     next();
-}
\ No newline at end of file
+  } catch (error) {
+    throw new ApiError(401, error?.message || "Invalid access token");
```
Comment (pavankpdev) [inline middlewares/auth.js]: here the status code should be 400

File: services/user/index.js [hunk 4]
```diff
@@ -53,7 +55,20 @@ export const findUserByEmail = async (email) => {
...
 
-export const findUserById = async (id) => { }
+/**
+ *
+ * @param {string} id - to unique identifier of the user
+ * @returns {Promise<userModel>} - to resolve to the find user
+ *  
+ */
+export const findUserById = async (id) => {
+  const user=await userModel.findById({
```
Comment (pavankpdev) [inline services/user/index.js]: you can pass the id directly, `findById(id)` no need to complicate it

**Review by pavankpdev — CHANGES_REQUESTED** [review]

File: services/user/index.js [hunk 5]
```diff
@@ -63,7 +63,7 @@ export const findUserByEmail = async (email) => {
  */
 export const findUserById = async (id) => {
   const user=await userModel.findById({
```
Comment (pavankpdev) [inline services/user/index.js]: You can pass the id directly `findById(id)`, here refer this [doc](https://mongoosejs.com/docs/api/model.html#Model.findById())

**Review by pavankpdev — CHANGES_REQUESTED** [review]
Hey @Sandeep1908 good job on the PR, just one more minor change required

File: services/user/index.js
Comment (pavankpdev) [inline services/user/index.js:1]: Great

**Review by pavankpdev — COMMENTED** [review]
Looks good @Sandeep1908 I'll test this change and then I'll merge it

**Comment by pavankpdev** [comment]
Hey @Sandeep1908 after testing the middleware I notices one major issue. 
The `APIError` reusable module we have, crashes the entire server instead of gracefully throwing error and sending it to the client. 

We can do this. 
1. We dispatch error response with status and message using `res` 
2. Or we set a global listener that checks for the error and responds the client with the error. 

You can choose any one of these.

**Comment by Sandeep1908** [comment]
@pavankpdev , please check I have used res for error handling

File: controllers/auth/index.js [hunk 6]
```diff
@@ -183,7 +182,7 @@ router.post("/login", async (req, res) => {
       .status(200);
   } catch (error) {
  
-    throw new ApiError(400, "ERRR in login");
+    return res.status(500).json({message: "ERRR in login"});
```
Comment (pavankpdev) [inline controllers/auth/index.js:185]: typo 😬

**Review by pavankpdev — APPROVED** [review]
@Sandeep1908 merging the PR, thanks for the contribution and your patience.

**Pull request merged**